instrumentation/aws-lambda/aws-lambda-events-2.2/library/src/main/java/io/opentelemetry/instrumentation/awslambdaevents/v2_2/TracingRequestWrapper.java

```diff
@@ -5,32 +5,90 @@
+import java.util.Map;
 
 /**
- * Wrapper for {@link io.opentelemetry.instrumentation.awslambdacore.v1_0.TracingRequestHandler}.
- * Allows for wrapping a regular lambda, not proxied through API Gateway. Therefore, HTTP headers
- * propagation is not supported.
+ * Wrapper for {@link com.amazonaws.services.lambda.runtime.RequestHandler} based Lambda handlers.
  */
-public class TracingRequestWrapper extends TracingRequestWrapperBase<Object, Object> {
+public class TracingRequestWrapper extends TracingRequestStreamWrapper {
   public TracingRequestWrapper() {
-    super(TracingRequestWrapper::map);
+    super();
   }
 
   // Visible for testing
-  TracingRequestWrapper(
-      OpenTelemetrySdk openTelemetrySdk,
-      WrappedLambda wrappedLambda,
-      BiFunction<Object, Class<?>, Object> mapper) {
-    super(openTelemetrySdk, wrappedLambda, mapper);
+  TracingRequestWrapper(OpenTelemetrySdk openTelemetrySdk, WrappedLambda wrappedLambda) {
+    super(openTelemetrySdk, wrappedLambda);
+  }
+
+  @Override
+  protected final AwsLambdaRequest createRequest(
+      InputStream inputStream, Context context, ApiGatewayProxyRequest proxyRequest) {
+    Method targetMethod = wrappedLambda.getRequestTargetMethod();
+    Object input = LambdaParameters.toInput(targetMethod, inputStream, TracingRequestWrapper::map);
+    return AwsLambdaRequest.create(context, input, extractHeaders(input));
+  }
+
+  protected Map<String, String> extractHeaders(Object input) {
+    if (input instanceof APIGatewayProxyRequestEvent) {
+      return MapUtils.emptyIfNull(((APIGatewayProxyRequestEvent) input).getHeaders());
+    }
+    return Collections.emptyMap();
+  }
+
+  @Override
+  protected final void doHandleRequest(
+      InputStream input, OutputStream output, Context context, AwsLambdaRequest request) {
+    Method targetMethod = wrappedLambda.getRequestTargetMethod();
+    Object[] parameters = LambdaParameters.toParameters(targetMethod, request.getInput(), context);
+    try {
+      Object result = targetMethod.invoke(wrappedLambda.getTargetObject(), parameters);
+      SerializationUtil.toJson(output, result);
+    } catch (IllegalAccessException e) {
+      throw new IllegalStateException("Method is inaccessible", e);
+    } catch (InvocationTargetException e) {
+      throw (e.getCause() instanceof RuntimeException
+          ? (RuntimeException) e.getCause()
+          : new IllegalStateException(e.getTargetException()));
+    }
+  }
+
+  @SuppressWarnings({"unchecked", "TypeParameterUnusedInFormals"})
+  // Used for testing
+  <I, O> O handleRequest(I input, Context context) throws IOException {
+    byte[] inputJsonData = SerializationUtil.toJsonData(input);
+    ByteArrayInputStream inputStream = new ByteArrayInputStream(inputJsonData);
+    ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+
+    super.handleRequest(inputStream, outputStream, context);
+
+    byte[] outputJsonData = outputStream.toByteArray();
+    return (O)
```

Comment:
Something visually strange about this to me. :) Can we just call these types `INPUT` and `OUTPUT` for clarity then?

Reply:
done